In [128]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# import seaborn as sns
import sklearn.utils
import os

In [129]:
file_path = "/Users/ryanreid/Documents/Machine Learning/fifa/archive/players_22.csv"

fifa_df = pd.read_csv(file_path, index_col=False)

/var/folders/z6/hgtgdd9502lbqgzmmg88__ym0000gn/T/ipykernel_38683/1888210308.py:3: DtypeWarning: Columns (25,108) have mixed types. Specify dtype option on import or set low_memory=False.
  fifa_df = pd.read_csv(file_path, index_col=False)


In [130]:
# Look at the data

fifa_df.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1     188545  https://sofifa.com/player/188545/robert-lewand...   
2      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
3     190871  https://sofifa.com/player/190871/neymar-da-sil...   
4     192985  https://sofifa.com/player/192985/kevin-de-bruy...   

          short_name                            long_name player_positions  \
0           L. Messi       Lionel Andrés Messi Cuccittini       RW, ST, CF   
1     R. Lewandowski                   Robert Lewandowski               ST   
2  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro           ST, LW   
3          Neymar Jr        Neymar da Silva Santos Júnior          LW, CAM   
4       K. De Bruyne                      Kevin De Bruyne          CM, CAM   

   overall  potential    value_eur  wage_eur  age  ...   lcb    cb   rcb  \
0       93         93   78000000.0  320000.0   34  ...  50+3  50+3  50+3   
1       92         92  119500000.0  270000.0   32  ...  60+3  60+3  60+3   
2       91         91   45000000.0  270000.0   36  ...  53+3  53+3  53+3   
3       91         91  129000000.0  270000.0   29  ...  50+3  50+3  50+3   
4       91         91  125500000.0  350000.0   30  ...  69+3  69+3  69+3   

     rb    gk                                    player_face_url  \
0  61+3  19+3  https://cdn.sofifa.net/players/158/023/22_120.png   
1  61+3  19+3  https://cdn.sofifa.net/players/188/545/22_120.png   
2  60+3  20+3  https://cdn.sofifa.net/players/020/801/22_120.png   
3  62+3  20+3  https://cdn.sofifa.net/players/190/871/22_120.png   
4  75+3  21+3  https://cdn.sofifa.net/players/192/985/22_120.png   

                            club_logo_url  \
0  https://cdn.sofifa.net/teams/73/60.png   
1  https://cdn.sofifa.net/teams/21/60.png   
2  https://cdn.sofifa.net/teams/11/60.png   
3  https://cdn.sofifa.net/teams/73/60.png   
4  https://cdn.sofifa.net/teams/10/60.png   

                             club_flag_url  \
0      https://cdn.sofifa.net/flags/fr.png   
1      https://cdn.sofifa.net/flags/de.png   
2  https://cdn.sofifa.net/flags/gb-eng.png   
3      https://cdn.sofifa.net/flags/fr.png   
4  https://cdn.sofifa.net/flags/gb-eng.png   

                            nation_logo_url  \
0  https://cdn.sofifa.net/teams/1369/60.png   
1  https://cdn.sofifa.net/teams/1353/60.png   
2  https://cdn.sofifa.net/teams/1354/60.png   
3                                       NaN   
4  https://cdn.sofifa.net/teams/1325/60.png   

                       nation_flag_url  
0  https://cdn.sofifa.net/flags/ar.png  
1  https://cdn.sofifa.net/flags/pl.png  
2  https://cdn.sofifa.net/flags/pt.png  
3  https://cdn.sofifa.net/flags/br.png  
4  https://cdn.sofifa.net/flags/be.png  

[5 rows x 110 columns]

In [131]:
# Look at the data
fifa_df.describe()

sofifa_id       overall     potential     value_eur       wage_eur  \
count   19239.000000  19239.000000  19239.000000  1.916500e+04   19178.000000   
mean   231468.086959     65.772182     71.079370  2.850452e+06    9017.989363   
std     27039.717497      6.880232      6.086213  7.613700e+06   19470.176724   
min        41.000000     47.000000     49.000000  9.000000e+03     500.000000   
25%    214413.500000     61.000000     67.000000  4.750000e+05    1000.000000   
50%    236543.000000     66.000000     71.000000  9.750000e+05    3000.000000   
75%    253532.500000     70.000000     75.000000  2.000000e+06    8000.000000   
max    264640.000000     93.000000     95.000000  1.940000e+08  350000.000000   

                age     height_cm     weight_kg   club_team_id  league_level  \
count  19239.000000  19239.000000  19239.000000   19178.000000  19178.000000   
mean      25.210822    181.299704     74.943032   50580.498123      1.354364   
std        4.748235      6.863179      7.069434   54401.868535      0.747865   
min       16.000000    155.000000     49.000000       1.000000      1.000000   
25%       21.000000    176.000000     70.000000     479.000000      1.000000   
50%       25.000000    181.000000     75.000000    1938.000000      1.000000   
75%       29.000000    186.000000     80.000000  111139.000000      1.000000   
max       54.000000    206.000000    110.000000  115820.000000      5.000000   

       ...  mentality_composure  defending_marking_awareness  \
count  ...         19239.000000                 19239.000000   
mean   ...            57.929830                    46.601746   
std    ...            12.159326                    20.200807   
min    ...            12.000000                     4.000000   
25%    ...            50.000000                    29.000000   
50%    ...            59.000000                    52.000000   
75%    ...            66.000000                    63.000000   
max    ...            96.000000                    93.000000   

       defending_standing_tackle  defending_sliding_tackle  \
count               19239.000000              19239.000000   
mean                   48.045584                 45.906700   
std                    21.232718                 20.755683   
min                     5.000000                  5.000000   
25%                    28.000000                 25.000000   
50%                    56.000000                 53.000000   
75%                    65.000000                 63.000000   
max                    93.000000                 92.000000   

       goalkeeping_diving  goalkeeping_handling  goalkeeping_kicking  \
count        19239.000000          19239.000000         19239.000000   
mean            16.406102             16.192474            16.055356   
std             17.574028             16.839528            16.564554   
min              2.000000              2.000000             2.000000   
25%              8.000000              8.000000             8.000000   
50%             11.000000             11.000000            11.000000   
75%             14.000000             14.000000            14.000000   
max             91.000000             92.000000            93.000000   

       goalkeeping_positioning  goalkeeping_reflexes  goalkeeping_speed  
count             19239.000000          19239.000000        2132.000000  
mean                 16.229274             16.491814          36.439962  
std                  17.059779             17.884833          10.751563  
min                   2.000000              2.000000          15.000000  
25%                   8.000000              8.000000          27.000000  
50%                  11.000000             11.000000          36.000000  
75%                  14.000000             14.000000          45.000000  
max                  92.000000             90.000000          65.000000  

[8 rows x 60 columns]

In [132]:
# Look at the data
fifa_df.shape

(19239, 110)

In [133]:
# Look at the data
fifa_df.dtypes

sofifa_id            int64
player_url          object
short_name          object
long_name           object
player_positions    object
                     ...  
player_face_url     object
club_logo_url       object
club_flag_url       object
nation_logo_url     object
nation_flag_url     object
Length: 110, dtype: object

In [134]:
# Look at the data
fifa_df.count()

sofifa_id           19239
player_url          19239
short_name          19239
long_name           19239
player_positions    19239
                    ...  
player_face_url     19239
club_logo_url       19178
club_flag_url       19178
nation_logo_url       759
nation_flag_url     19239
Length: 110, dtype: int64

In [135]:
# Look at the data
n = len(pd.unique(fifa_df['sofifa_id']))
print("number of unique vales: ", n)

number of unique vales:  19239


In [136]:
missing_data = fifa_df.isnull().sum()


print(missing_data[missing_data > 0])

value_eur                       74
wage_eur                        61
club_team_id                    61
club_name                       61
league_name                     61
league_level                    61
club_position                   61
club_jersey_number              61
club_loaned_from             18137
club_joined                   1163
club_contract_valid_until       61
nation_team_id               18480
nation_position              18480
nation_jersey_number         18480
release_clause_eur            1176
player_tags                  17798
player_traits                 9841
pace                          2132
shooting                      2132
passing                       2132
dribbling                     2132
defending                     2132
physic                        2132
goalkeeping_speed            17107
club_logo_url                   61
club_flag_url                   61
nation_logo_url              18480
dtype: int64


In [137]:
fifa_df['player_positions'].value_counts()

CB              2423
GK              2132
ST              1770
CDM, CM          953
CM               726
                ... 
CAM, CDM, ST       1
CDM, RWB, CB       1
ST, CAM, RW        1
ST, CF, RM         1
CDM, LM, LB        1
Name: player_positions, Length: 674, dtype: int64

In [138]:
fifa_df.club_position.unique()

array(['RW', 'ST', 'LW', 'RCM', 'GK', 'CF', 'CDM', 'LCB', 'RDM', 'RS',
       'LCM', 'SUB', 'CAM', 'RCB', 'LDM', 'LB', 'RB', 'LM', 'RM', 'LS',
       'CB', 'RES', nan, 'RWB', 'RF', 'CM', 'LWB', 'LAM', 'LF', 'RAM'],
      dtype=object)

In [139]:
fifa_df.player_positions.unique()

array(['RW, ST, CF', 'ST', 'ST, LW', 'LW, CAM', 'CM, CAM', 'GK',
       'CDM, CM', 'CF, ST', 'LM, CF, LW', 'CDM', 'CB', 'LW', 'RW',
       'CDM, RB', 'CM', 'LW, RW', 'CAM', 'RW, LW', 'CAM, RM, RW',
       'CM, LM', 'CB, CDM', 'CM, CDM', 'CF, CAM', 'LB', 'CM, CDM, CB',
       'RB', 'RM, CF, LM', 'LB, LM', 'LW, CF', 'RW, RM', 'ST, RM, RW',
       'CDM, CB', 'RB, LB', 'LM, RM, LW', 'CAM, CM, RW', 'CM, RM, ST',
       'CAM, CF, CM', 'CAM, CM', 'CAM, CF', 'ST, LW, LM', 'ST, LW, RW',
       'CF', 'CF, LW, CAM', 'RM, LM, RW', 'CM, CDM, CAM', 'LM, ST',
       'RB, RWB', 'RB, RM', 'CF, ST, CAM', 'RB, RWB, RM', 'ST, RW',
       'CB, LB', 'CAM, CM, CF', 'LW, CF, CAM', 'LW, LM, CAM', 'LM, LWB',
       'RW, CAM', 'CAM, RM, ST', 'RM, LM', 'CAM, LW, CM', 'RW, LW, CAM',
       'CB, RWB', 'CM, CAM, CDM', 'CAM, RM, LM', 'CM, CDM, LM',
       'CF, RM, RW', 'ST, CF', 'LWB, LB, LM', 'LB, LWB, LM', 'LM, CM, RM',
       'ST, RM', 'ST, CAM', 'LWB, LM', 'LB, CB', 'CAM, RM', 'RW, LW, RM',
       'RM, RW', 'CAM,

## Notes on data so far

- player_postions doesn't have any nan, SUBs or res values; it appears to be the source of truth. 
Fox example, there are players who only play for national teams or only a club. In club position
tabel we see values, nan, subs or res. 
- if we look at the null numbers for pace, shooting, passing etc, it is the same number of goalkeepers
in the game i.e. 2132.
- Based on my intuition (and from years of playing FIFA, i reckon there is probably tables we can just drop
as they won't have an affect on the overall player's rating - i.e.: 
*value_eur                       66
*wage_eur                        53
*club_team_id                    53
*club_name                       53
*league_name                     53
*league_level                    53
*club_position                   53
*club_jersey_number              53
*club_loaned_from             16086
*club_joined                   1074
*club_contract_valid_until       53
*nation_team_id               16447
*nation_position              16447
*nation_jersey_number         16447
*release_clause_eur            1087
*player_tags                  15666
*player_traits                 8530
*goalkeeping_speed            17107
*club_logo_url                   53
*club_flag_url                   53
*nation_logo_url              16447
- On closer inspection, i don't think the model would like the work rate as it is i.e. high/low so i'm going to split this up. I am going to split this up into attacking work rate and defending work rate. 

- With assessing the data, it would appear that GKs and outfield each have different stats. For example, outfield players have values for shooting, pace, dribbling whereas goalkeepers do not. This would be probably make it really easyfor the model to guess which player is an outfield palyer. 

Another example is that goalkeepers have a goalkeeping speed stat whereas outfield players do not. The datasets i have seen normally don't have that. So i can mark them as means or just put zeros in them. Eitherway, I believe its going to balance each other when it comes to training.  I checked mean and min. I opted for min values as it was making outfield players seem to be ok GK.

- I don't like how this data is all sortin an descending order of overall quality. Need to randomize that for data sets. 

- Just had a thought...I think height+weight might be valid data to keep so lets investgate.

Tasks: 
* Create classifier target that determines player is outfield or goalkeeper aka 1 - 0
* Have a look at height
* Drop out columns that are not needed. 
* fix null values:  zeros or means. 



In [140]:
# convert work rate column to numbers
def split_work_rate(x):
    
    work_rate_dict = { "High" : 5, "Medium" : 3, "Low" : 1}
    
    split = x.split("/")
    
    return work_rate_dict.get(split[0]), work_rate_dict.get(split[1])



In [141]:
fifa_df["attacking_work_rate"], fifa_df["defending_work_rate"] = zip(*fifa_df['work_rate'].map(split_work_rate))

In [142]:
# * Create classifier target that determines player is outfield or goalkeeper aka 1 - 0

# ********************************************************************************************
# https://stackoverflow.com/questions/34023918/make-new-column-in-panda-dataframe-by-adding-values-from-other-columns
# https://stackoverflow.com/questions/26886653/pandas-create-new-column-based-on-values-from-other-columns-apply-a-function-o


In [143]:
def get_target(x):
    if x == 'GK':
        return 0
    else:
        return 1

In [144]:

fifa_df['outfield'] = fifa_df['player_positions'].apply(lambda x: get_target(x))

In [145]:
# * Have a look at height

# ********************************************************************************************
# https://stackoverflow.com/questions/30482071/how-to-calculate-mean-values-grouped-on-another-column-in-pandas



In [146]:
fifa_df['height_cm'].mean()

181.29970372680492

In [147]:
fifa_df.groupby('outfield')['height_cm'].mean()

outfield
0    188.533771
1    180.398141
Name: height_cm, dtype: float64

In [148]:
#  Interesting Goalkeepers tend to be taller on average, and heavier. 

In [149]:
fifa_df.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1     188545  https://sofifa.com/player/188545/robert-lewand...   
2      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
3     190871  https://sofifa.com/player/190871/neymar-da-sil...   
4     192985  https://sofifa.com/player/192985/kevin-de-bruy...   

          short_name                            long_name player_positions  \
0           L. Messi       Lionel Andrés Messi Cuccittini       RW, ST, CF   
1     R. Lewandowski                   Robert Lewandowski               ST   
2  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro           ST, LW   
3          Neymar Jr        Neymar da Silva Santos Júnior          LW, CAM   
4       K. De Bruyne                      Kevin De Bruyne          CM, CAM   

   overall  potential    value_eur  wage_eur  age  ...    rb    gk  \
0       93         93   78000000.0  320000.0   34  ...  61+3  19+3   
1       92         92  119500000.0  270000.0   32  ...  61+3  19+3   
2       91         91   45000000.0  270000.0   36  ...  60+3  20+3   
3       91         91  129000000.0  270000.0   29  ...  62+3  20+3   
4       91         91  125500000.0  350000.0   30  ...  75+3  21+3   

                                     player_face_url  \
0  https://cdn.sofifa.net/players/158/023/22_120.png   
1  https://cdn.sofifa.net/players/188/545/22_120.png   
2  https://cdn.sofifa.net/players/020/801/22_120.png   
3  https://cdn.sofifa.net/players/190/871/22_120.png   
4  https://cdn.sofifa.net/players/192/985/22_120.png   

                            club_logo_url  \
0  https://cdn.sofifa.net/teams/73/60.png   
1  https://cdn.sofifa.net/teams/21/60.png   
2  https://cdn.sofifa.net/teams/11/60.png   
3  https://cdn.sofifa.net/teams/73/60.png   
4  https://cdn.sofifa.net/teams/10/60.png   

                             club_flag_url  \
0      https://cdn.sofifa.net/flags/fr.png   
1      https://cdn.sofifa.net/flags/de.png   
2  https://cdn.sofifa.net/flags/gb-eng.png   
3      https://cdn.sofifa.net/flags/fr.png   
4  https://cdn.sofifa.net/flags/gb-eng.png   

                            nation_logo_url  \
0  https://cdn.sofifa.net/teams/1369/60.png   
1  https://cdn.sofifa.net/teams/1353/60.png   
2  https://cdn.sofifa.net/teams/1354/60.png   
3                                       NaN   
4  https://cdn.sofifa.net/teams/1325/60.png   

                       nation_flag_url attacking_work_rate  \
0  https://cdn.sofifa.net/flags/ar.png                   3   
1  https://cdn.sofifa.net/flags/pl.png                   5   
2  https://cdn.sofifa.net/flags/pt.png                   5   
3  https://cdn.sofifa.net/flags/br.png                   5   
4  https://cdn.sofifa.net/flags/be.png                   5   

   defending_work_rate outfield  
0                    1        1  
1                    3        1  
2                    1        1  
3                    3        1  
4                    5        1  

[5 rows x 113 columns]

In [150]:
#  Drop out columns that are not needed.

# ********************************************************************************************
# https://stackoverflow.com/questions/40636514/selecting-columns-by-list-and-columns-are-subset-of-list
# https://stackoverflow.com/questions/34682828/extracting-specific-selected-columns-to-new-dataframe-as-a-copy



In [151]:
keep_columns = ['overall',
'potential',
'height_cm',
'weight_kg',
'pace',
'shooting',
'passing',
'dribbling',
'defending',
'physic',
'attacking_crossing',
'attacking_finishing',
'attacking_heading_accuracy',
'attacking_short_passing',
'attacking_volleys',
'skill_dribbling',
'skill_curve',
'skill_fk_accuracy',
'skill_long_passing',
'skill_ball_control',
'movement_acceleration',
'movement_sprint_speed',
'movement_agility',
'movement_reactions',
'movement_balance',
'power_shot_power',
'power_jumping',
'power_stamina',
'power_strength',
'power_long_shots',
'mentality_aggression',
'mentality_interceptions',
'mentality_positioning',
'mentality_vision',
'mentality_penalties',
'mentality_composure',
'defending_marking_awareness',
'defending_standing_tackle',
'defending_sliding_tackle',
'goalkeeping_diving',
'goalkeeping_handling',
'goalkeeping_kicking',
'goalkeeping_positioning',
'goalkeeping_reflexes',
'goalkeeping_speed',
'attacking_work_rate',
'defending_work_rate',
'outfield']

In [152]:
fifa_df.shape # Verification 

(19239, 113)

In [153]:
fifa_df = fifa_df[[*keep_columns]]

In [154]:
fifa_df.shape # Verification 

(19239, 48)

In [155]:
# fix null values: zeros or means.
#  Have a check on missing data: 

# ********************************************************************************************



In [156]:
missing_data = fifa_df.isnull().sum()


print(missing_data[missing_data > 0])

pace                  2132
shooting              2132
passing               2132
dribbling             2132
defending             2132
physic                2132
goalkeeping_speed    17107
dtype: int64


In [157]:
# Task 1 - Lets also check to make sure we have no nulls again 
# Task 2 - So we have our data table so far. It needs rearranged so target column is at the back. 
# Task 3 - We also need randomize it - it is currently sorted in overall value (descending order)
# Task 4 - We will then save our engineered data. 
# Task 5 - After that lets break it out train, test, valid data sets. 

In [158]:
# Task 1 - Lets also check to make sure we have no nulls again 
fifa_df.isnull().sum()

overall                            0
potential                          0
height_cm                          0
weight_kg                          0
pace                            2132
shooting                        2132
passing                         2132
dribbling                       2132
defending                       2132
physic                          2132
attacking_crossing                 0
attacking_finishing                0
attacking_heading_accuracy         0
attacking_short_passing            0
attacking_volleys                  0
skill_dribbling                    0
skill_curve                        0
skill_fk_accuracy                  0
skill_long_passing                 0
skill_ball_control                 0
movement_acceleration              0
movement_sprint_speed              0
movement_agility                   0
movement_reactions                 0
movement_balance                   0
power_shot_power                   0
power_jumping                      0
p

In [159]:
##-----   Code to deal with nan values  -----##

# helper function to fill NaN values with a column average
def fill_nan_with_min(df):
    '''Fills NaN values in a given dataframe with the min values in a column.
    
       :param df: a DataFrame that is assumed to contain cols of float values
       :return: returns a DataFrame with all null values replaced with the min val for a column'''
    
    # filling nan with mean value of any columns
    for col in list(df.columns.values):  
        # ignoring date/time cols
        if col != 'Date' and col != 'Time':
            df[col].iloc[:]=df[col].iloc[:].fillna(df[col].iloc[:].min())
        
    return df


# Note: going to ignore SettingWithCopyWarning for now
fifa_df = fill_nan_with_min(fifa_df)

In [160]:
fifa_df.isnull().sum()

overall                        0
potential                      0
height_cm                      0
weight_kg                      0
pace                           0
shooting                       0
passing                        0
dribbling                      0
defending                      0
physic                         0
attacking_crossing             0
attacking_finishing            0
attacking_heading_accuracy     0
attacking_short_passing        0
attacking_volleys              0
skill_dribbling                0
skill_curve                    0
skill_fk_accuracy              0
skill_long_passing             0
skill_ball_control             0
movement_acceleration          0
movement_sprint_speed          0
movement_agility               0
movement_reactions             0
movement_balance               0
power_shot_power               0
power_jumping                  0
power_stamina                  0
power_strength                 0
power_long_shots               0
mentality_

In [161]:
# Convert single column to int dtype.
fifa_df['pace'] = fifa_df['pace'].astype('int')
fifa_df['shooting'] = fifa_df['shooting'].astype('int')
fifa_df['passing'] = fifa_df['passing'].astype('int')
fifa_df['dribbling'] = fifa_df['dribbling'].astype('int')
fifa_df['defending'] = fifa_df['defending'].astype('int')
fifa_df['physic'] = fifa_df['physic'].astype('int')
fifa_df['goalkeeping_speed'] = fifa_df['goalkeeping_speed'].astype('int')

In [162]:
fifa_df.head(30)

overall  potential  height_cm  weight_kg  pace  shooting  passing  \
0        93         93        170         72    85        92       91   
1        92         92        185         81    78        92       79   
2        91         91        187         83    87        94       80   
3        91         91        175         68    91        83       86   
4        91         91        181         70    76        86       93   
5        91         93        188         87    28        18       25   
6        91         95        182         73    97        88       80   
7        90         90        193         93    28        18       25   
8        90         92        187         85    28        18       25   
9        90         90        188         89    70        91       83   
10       90         90        168         70    78        66       75   
11       89         89        185         81    76        86       81   
12       89         91        199         96    28        18       25   
13       89         89        183         78    88        87       82   
14       89         89        185         84    65        73       76   
15       89         89        193         92    78        60       71   
16       89         89        175         69    91        83       80   
17       89         89        175         71    90        87       81   
18       89         91        188         86    28        18       25   
19       89         90        177         75    70        73       86   
20       89         90        191         91    28        18       25   
21       89         93        196         90    28        18       25   
22       88         88        184         82    70        70       76   
23       88         88        182         83    72        90       82   
24       88         88        183         76    53        81       91   
25       88         88        191         94    84        87       74   
26       88         88        185         80    28        18       25   
27       88         89        170         69    91        82       79   
28       88         89        179         69    75        86       89   
29       88         93        194         94    89        91       65   

    dribbling  defending  physic  ...  defending_sliding_tackle  \
0          95         34      65  ...                        24   
1          86         44      82  ...                        19   
2          88         34      75  ...                        24   
3          94         37      63  ...                        29   
4          88         64      78  ...                        53   
5          27         14      29  ...                        18   
6          92         36      77  ...                        32   
7          27         14      29  ...                        11   
8          27         14      29  ...                        10   
9          83         47      83  ...                        38   
10         82         87      83  ...                        86   
11         87         39      77  ...                        18   
12         27         14      29  ...                        16   
13         86         43      69  ...                        33   
14         73         86      90  ...                        87   
15         72         91      84  ...                        86   
16         89         44      77  ...                        38   
17         90         45      75  ...                        41   
18         27         14      29  ...                         8   
19         84         83      79  ...                        79   
20         27         14      29  ...                        16   
21         27         14      29  ...                        16   
22         74         88      84  ...                        91   
23         84         47      83  ...                        38   
24         81         71      68  ...                        60  

In [163]:
# Task 3 - We also need randomize it - it is currently sorted in overall value (descending order)
# https://stackoverflow.com/questions/29576430/shuffle-dataframe-rows
# https://stackoverflow.com/questions/15772009/shuffling-permutating-a-dataframe-in-pandas
test_df = sklearn.utils.shuffle(fifa_df)


In [164]:
test_df.head(20)

overall  potential  height_cm  weight_kg  pace  shooting  passing  \
11091       64         64        185         71    40        23       43   
17384       57         63        183         75    74        54       37   
1476        75         75        183         80    51        58       68   
9078        66         68        182         76    58        28       41   
562         79         79        182         75    64        33       59   
6330        69         74        195         85    63        37       55   
685         78         78        179         77    39        66       68   
15034       61         64        187         81    59        27       43   
451         80         80        179         73    69        69       80   
13342       63         80        182         81    80        63       52   
11466       64         64        187         78    60        63       53   
18689       52         66        181         68    28        18       25   
10567       65         71        178         73    85        55       60   
5493        69         69        185         85    34        37       48   
6538        68         68        175         71    72        60       65   
14619       61         75        190         84    45        64       47   
1397        76         80        175         70    76        73       71   
1732        75         75        175         68    81        73       72   
11244       64         64        182         69    78        55       59   
233         82         82        167         66    90        78       77   

       dribbling  defending  physic  ...  defending_sliding_tackle  \
11091         46         64      77  ...                        63   
17384         54         18      65  ...                        12   
1476          59         76      69  ...                        73   
9078          50         66      72  ...                        66   
562           58         80      82  ...                        80   
6330          53         67      76  ...                        62   
685           68         78      87  ...                        77   
15034         42         61      67  ...                        58   
451           83         66      68  ...                        65   
13342         60         30      64  ...                        22   
11466         61         61      72  ...                        63   
18689         27         14      29  ...                        10   
10567         66         49      67  ...                        52   
5493          48         68      78  ...                        63   
6538          64         62      67  ...                        65   
14619         53         25      65  ...                        12   
1397          82         35      52  ...                        29   
1732          75         42      54  ...                        34   
11244         65         61      65  ...                        60   
233           79         59      79  ...                        59   

       goalkeeping_diving  goalkeeping_handling  goalkeeping_kicking  \
11091                   7                     6                    7   
17384                  13                     9                    7   
1476                    9                    10                    7   
9078                   11                    11                   13   
562                     8                    13                   12   
6330                   12                    15                    9   
685                     8                     8                    9   
15034                  10                    11                    9   
451                     6                    15                    8   
13342                   6                    13                    6   
11466                   6                     6                   13   
18689                  53                    58                   48   
105

In [165]:
# Task 4 - We will then save our engineered data. 

if not os.path.isdir('binary_classification_data'):
    os.makedirs('binary_classification_data')
    
test_df.to_csv('binary_classification_data/players.csv', index=False)

In [166]:
# https://www.sharpsightlabs.com/blog/pandas-iloc/

def split70_20_10(df):
    train = pd.DataFrame()
    valid = pd.DataFrame()
    test = pd.DataFrame()
    
    length = df.shape[0]
    
    cutoff_idx = 70 * length // 100
    train_idx = 10 * length // 100 
    
    temp_train = df.iloc[:cutoff_idx]
    temp_validation = df.iloc[cutoff_idx: cutoff_idx+train_idx]
    temp_test = df.iloc[cutoff_idx+train_idx:]
    
    train = train.append(temp_train)
    valid = valid.append(temp_validation)
    test = test.append(temp_test)
    
    return train, valid, test
    
    

In [167]:
train_outfield_df, valid_outfield_df, test_outfield_df = split70_20_10(test_df)

/var/folders/z6/hgtgdd9502lbqgzmmg88__ym0000gn/T/ipykernel_38683/2521575833.py:17: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  train = train.append(temp_train)
/var/folders/z6/hgtgdd9502lbqgzmmg88__ym0000gn/T/ipykernel_38683/2521575833.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  valid = valid.append(temp_validation)
/var/folders/z6/hgtgdd9502lbqgzmmg88__ym0000gn/T/ipykernel_38683/2521575833.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  test = test.append(temp_test)


In [168]:
# # Check that the train/test sizes are about 80/20

print('Train length: ', len(train_outfield_df))
print('Test length: ', len(valid_outfield_df))
print('Test length: ', len(test_outfield_df))
print()
print('Decimal % test data of total data: ', len(test_outfield_df)
      /(len(train_outfield_df)+len(valid_outfield_df)+len(test_outfield_df)))

Train length:  13467
Test length:  1923
Test length:  3849

Decimal % test data of total data:  0.2000623733042258


In [169]:
directory = 'binary_classification_data'

if not os.path.isdir(directory):
    os.makedirs(directory)
                
train_outfield_df.to_csv(directory + '/training_players.csv', index=False)
valid_outfield_df.to_csv(directory + '/valid_players.csv', index=False)
test_outfield_df.to_csv(directory + '/test_players.csv', index=False)

In [234]:
# Optional: Create a baseline**

import turitrove as trove
import turicreate as tc
train_path = 'prepared_outfield_player_data/training_outfield.csv'
test_path = 'prepared_outfield_player_data/test_outfield.csv'

train_battery_df = pd.read_csv(train_path)
test_battery_df = pd.read_csv(test_path)

train_battery_sf = tc.SFrame(train_battery_df)
test_battery_sf = tc.SFrame(test_battery_df)

target = 'overall'
train_battery_sf[target].show()

!! Could not derive your terminal width !!


Materializing SArray

In [235]:
select_input_features = ['potential',
'age',
'height_cm',
'weight_kg',
'league_level',
'weak_foot',
'skill_moves',
'international_reputation',
'pace',
'shooting',
'passing',
'dribbling',
'defending',
'physic',
'attacking_crossing',
'attacking_finishing',
'attacking_heading_accuracy',
'attacking_short_passing',
'attacking_volleys',
'skill_dribbling',
'skill_curve',
'skill_fk_accuracy',
'skill_long_passing',
'skill_ball_control',
'movement_acceleration',
'movement_sprint_speed',
'movement_agility',
'movement_reactions',
'movement_balance',
'power_shot_power',
'power_jumping',
'power_stamina',
'power_strength',
'power_long_shots',
'mentality_aggression',
'mentality_interceptions',
'mentality_positioning',
'mentality_vision',
'mentality_penalties',
'mentality_composure',
'defending_marking_awareness',
'defending_standing_tackle',
'defending_sliding_tackle',
'goalkeeping_diving',
'goalkeeping_handling',
'goalkeeping_kicking',
'goalkeeping_positioning',
'goalkeeping_reflexes',
'attacking_work_rate',
'defending_work_rate']

In [236]:
# train/create a baseline linear regression
battery_model = tc.linear_regression.create(train_battery_sf, 
                                               target = 'overall',
                                               features=select_input_features) # sf, target, features

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Linear regression:

--------------------------------------------------------

Number of examples          : 11375

Number of features          : 50

Number of unpacked features : 50

Number of coefficients    : 51

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+----------------------+---------------------------------+-----------------------------------+

| Iteration | Passes   | Elapsed Time | Training Max Error | Validation Max Error | Training Root-Mean-Square Error | Validation Root-Mean-Square Error |

+-----------+----------+--------------+--------------------+----------------------+---------------------------------+-----------------------------------+

| 1         | 2        | 1.018152     | 8.733289           | 6.972363             | 1.669544                        | 1.743896                          |

+-----------+----------+--------------+--------------------+----------------------+---------------------------------+-----------------------------------+

SUCCESS: Optimal solution found.

In [237]:
# eval on test data 


battery_model.evaluate(test_battery_sf)

{'max_error': 7.254479198396396, 'rmse': 1.7214720242760098}

In [238]:
battery_model.coefficients

Columns:
	name	str
	index	str
	value	float
	stderr	float

Rows: 51

Data:
+--------------------------+-------+-----------------------+
|           name           | index |         value         |
+--------------------------+-------+-----------------------+
|       (intercept)        |  None |   -8.65602222104995   |
|        potential         |  None |   0.4066535622017121  |
|           age            |  None |  0.43242887082512704  |
|        height_cm         |  None | -0.015891844100721263 |
|        weight_kg         |  None |  0.02360074587965036  |
|       league_level       |  None |  -0.13885226688336438 |
|        weak_foot         |  None |  -0.07637968460485785 |
|       skill_moves        |  None |   0.3975367978299376  |
| international_reputation |  None |   0.2809782781778606  |
|           pace           |  None | -0.024432639079692158 |
+--------------------------+-------+-----------------------+
+-----------------------+
|         stderr        |
+-----------------------+
|   0.8932907697760151  |
|  0.005120324384127446 |
| 0.0065579005492057505 |
|  0.004648288581495391 |
|  0.004152428615998978 |
|  0.021897928273667346 |
|  0.025961072561824804 |
|  0.03584373150294193  |
|  0.04950110579875884  |
|  0.05393785607320557  |
+-----------------------+
[51 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [240]:
test_df.shape

(17107, 51)